File: crates/moss-core/src/lex.rs
use std::{
    fmt,
    ops::{Add, Range},
};

use enumset::EnumSetType;
use index_vec::{IndexVec, define_index_type};
use logos::Logos;

define_index_type! {
    pub struct ByteIndex = u32;
    IMPL_RAW_CONVERSIONS = true;
}

define_index_type! {
    pub struct TokenId = u32;
}

impl Add<i32> for TokenId {
    type Output = Self;

    fn add(self, rhs: i32) -> Self::Output {
        Self::from_raw(if rhs < 0 {
            self.raw() - (-rhs) as u32
        } else {
            self.raw() + rhs as u32
        })
    }
}

#[derive(Debug, EnumSetType, Logos)]
#[logos(skip r"(#[^\n]+|\s)+")]
pub enum Token {
    Eof,

    #[token("!")]
    Exclam,

    #[token("%")]
    Percent,

    #[token("(")]
    LParen,

    #[token(")")]
    RParen,

    #[token("*")]
    Star,

    #[token("+")]
    Plus,

    #[token(",")]
    Comma,

    #[token("-")]
    Hyphen,

    #[token(".")]
    Dot,

    #[token("/")]
    Slash,

    #[token(":")]
    Colon,

    #[token(";")]
    Semi,

    #[token("<")]
    Less,

    #[token("=")]
    Equal,

    #[token(">")]
    Greater,

    #[token("[")]
    LBracket,

    #[token("]")]
    RBracket,

    #[token("{")]
    LBrace,

    #[token("}")]
    RBrace,

    #[token("!=")]
    ExclamEqual,

    #[token("::")]
    ColonColon,

    #[token("<=")]
    LessEqual,

    #[token("==")]
    EqualEqual,

    #[token(">=")]
    GreaterEqual,

    #[token("as")]
    As,

    #[token("assume")]
    Assume,

    #[token("context")]
    Context,

    #[token("else")]
    Else,

    #[token("fn")]
    Fn,

    #[token("if")]
    If,

    #[token("import")]
    Import,

    #[token("let")]
    Let,

    #[token("static")]
    Static,

    #[token("struct")]
    Struct,

    #[token("this")]
    This,

    #[token("type")]
    Type,

    #[token("use")]
    Use,

    #[token("val")]
    Val,

    #[token("var")]
    Var,

    #[token("while")]
    While,

    #[regex(r"[A-Z_a-z]\w*")]
    Name,

    #[regex(r#""([^"\\\n]|\\["\\nrt])*""#)]
    Str,

    #[regex(r"\d+")]
    Int,
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Token::Eof => write!(f, "end of file"),
            Token::Exclam => write!(f, "`!`"),
            Token::Percent => write!(f, "`%`"),
            Token::LParen => write!(f, "`(`"),
            Token::RParen => write!(f, "`)`"),
            Token::Star => write!(f, "`*`"),
            Token::Plus => write!(f, "`+`"),
            Token::Comma => write!(f, "`,`"),
            Token::Hyphen => write!(f, "`-`"),
            Token::Dot => write!(f, "`.`"),
            Token::Slash => write!(f, "`/`"),
            Token::Colon => write!(f, "`:`"),
            Token::Semi => write!(f, "`;`"),
            Token::Less => write!(f, "`<`"),
            Token::Equal => write!(f, "`=`"),
            Token::Greater => write!(f, "`>`"),
            Token::LBracket => write!(f, "`[`"),
            Token::RBracket => write!(f, "`]`"),
            Token::LBrace => write!(f, "`{{`"),
            Token::RBrace => write!(f, "`}}`"),
            Token::ExclamEqual => write!(f, "`!=`"),
            Token::ColonColon => write!(f, "`::`"),
            Token::LessEqual => write!(f, "`<=`"),
            Token::EqualEqual => write!(f, "`==`"),
            Token::GreaterEqual => write!(f, "`>=`"),
            Token::As => write!(f, "`as`"),
            Token::Assume => write!(f, "`assume`"),
            Token::Context => write!(f, "`context`"),
            Token::Else => write!(f, "`else`"),
            Token::Fn => write!(f, "`fn`"),
            Token::If => write!(f, "`if`"),
            Token::Import => write!(f, "`import`"),
            Token::Let => write!(f, "`let`"),
            Token::Static => write!(f, "`static`"),
            Token::Struct => write!(f, "`struct`"),
            Token::This => write!(f, "`this`"),
            Token::Type => write!(f, "`type`"),
            Token::Use => write!(f, "`use`"),
            Token::Val => write!(f, "`val`"),
            Token::Var => write!(f, "`var`"),
            Token::While => write!(f, "`while`"),
            Token::Name => write!(f, "name"),
            Token::Str => write!(f, "string"),
            Token::Int => write!(f, "integer"),
        }
    }
}

pub type Tokens = IndexVec<TokenId, Token>;

pub type TokenStarts = IndexVec<TokenId, ByteIndex>;

#[derive(Clone, Copy, Debug)]
pub enum LexError {
    SourceTooLong,
    InvalidToken { start: ByteIndex, end: ByteIndex },
}

impl LexError {
    pub fn byte_range(self) -> Range<usize> {
        match self {
            LexError::SourceTooLong => {
                let max = ByteIndex::from(u32::MAX).index();
                max..max
            }
            LexError::InvalidToken { start, end } => start.index()..end.index(),
        }
    }

    pub fn message(self) -> &'static str {
        match self {
            LexError::SourceTooLong => "file size exceeds 4 GiB limit",
            LexError::InvalidToken { .. } => "invalid token",
        }
    }
}

pub fn lex(source: &str) -> Result<(Tokens, TokenStarts), LexError> {
    let Ok(eof) = u32::try_from(source.len()) else {
        return Err(LexError::SourceTooLong);
    };
    let mut tokens = IndexVec::new();
    let mut starts = IndexVec::new();
    for (result, range) in Token::lexer(source).spanned() {
        let start = ByteIndex::from_usize(range.start);
        let end = ByteIndex::from_usize(range.end);
        let token = result.map_err(|_| LexError::InvalidToken { start, end })?;
        tokens.push(token);
        starts.push(start);
    }
    tokens.push(Token::Eof);
    starts.push(ByteIndex::from(eof));
    Ok((tokens, starts))
}

pub fn relex(source: &str, starts: &TokenStarts, token: TokenId) -> Range<usize> {
    let start = starts[token].index();
    let (_, range) = Token::lexer(&source[start..]).spanned().next().unwrap();
    (range.start + start)..(range.end + start)
}

File: bootstrap/mossc/lex.py
"""Lexer for the MVP token set of docs/reference/syntax.md."""

from dataclasses import dataclass
from enum import Enum, auto


class Kind(Enum):
    # One-character symbols.
    EXCLAM = auto()
    PERCENT = auto()
    AMPERSAND = auto()
    LPAREN = auto()
    RPAREN = auto()
    STAR = auto()
    PLUS = auto()
    COMMA = auto()
    HYPHEN = auto()
    HYPHEN_GREATER = auto()
    DOT = auto()
    SLASH = auto()
    COLON = auto()
    SEMI = auto()
    LESS = auto()
    EQUAL = auto()
    GREATER = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    CARET = auto()
    LBRACE = auto()
    PIPE = auto()
    RBRACE = auto()
    # Two-character symbols.
    EXCLAM_EQUAL = auto()
    COLON_COLON = auto()
    EQUAL_GREATER = auto()
    LESS_LESS = auto()
    LESS_EQUAL = auto()
    EQUAL_EQUAL = auto()
    GREATER_EQUAL = auto()
    GREATER_GREATER = auto()
    # Keywords.
    AS = auto()
    ASSUME = auto()
    BIND = auto()
    BREAK = auto()
    CONTEXT = auto()
    ELSE = auto()
    FN = auto()
    FOR = auto()
    FUNCTOR = auto()
    IF = auto()
    IMPORT = auto()
    LET = auto()
    LOOP = auto()
    MATCH = auto()
    RETURN = auto()
    THIS_TYPE = auto()  # This
    THIS = auto()  # this
    TYPE = auto()
    UNIT = auto()
    USE = auto()
    VAL = auto()
    VAR = auto()
    WHILE = auto()
    # Everything else.
    NAME = auto()
    STRING = auto()
    EOF = auto()


KEYWORDS = {
    "as": Kind.AS,
    "assume": Kind.ASSUME,
    "functor": Kind.FUNCTOR,
    "bind": Kind.BIND,
    "break": Kind.BREAK,
    "context": Kind.CONTEXT,
    "else": Kind.ELSE,
    "fn": Kind.FN,
    "for": Kind.FOR,
    "if": Kind.IF,
    "import": Kind.IMPORT,
    "let": Kind.LET,
    "loop": Kind.LOOP,
    "match": Kind.MATCH,
    "return": Kind.RETURN,
    "This": Kind.THIS_TYPE,
    "this": Kind.THIS,
    "type": Kind.TYPE,
    "unit": Kind.UNIT,
    "use": Kind.USE,
    "val": Kind.VAL,
    "var": Kind.VAR,
    "while": Kind.WHILE,
}

ONE_CHAR = {
    "!": Kind.EXCLAM,
    "%": Kind.PERCENT,
    "&": Kind.AMPERSAND,
    "(": Kind.LPAREN,
    ")": Kind.RPAREN,
    "*": Kind.STAR,
    "+": Kind.PLUS,
    ",": Kind.COMMA,
    "-": Kind.HYPHEN,
    ".": Kind.DOT,
    "/": Kind.SLASH,
    ":": Kind.COLON,
    ";": Kind.SEMI,
    "<": Kind.LESS,
    "=": Kind.EQUAL,
    ">": Kind.GREATER,
    "[": Kind.LBRACKET,
    "]": Kind.RBRACKET,
    "^": Kind.CARET,
    "{": Kind.LBRACE,
    "|": Kind.PIPE,
    "}": Kind.RBRACE,
}

TWO_CHAR = {
    "!=": Kind.EXCLAM_EQUAL,
    "::": Kind.COLON_COLON,
    "=>": Kind.EQUAL_GREATER,
    "<<": Kind.LESS_LESS,
    "<=": Kind.LESS_EQUAL,
    "==": Kind.EQUAL_EQUAL,
    ">=": Kind.GREATER_EQUAL,
    ">>": Kind.GREATER_GREATER,
    "->": Kind.HYPHEN_GREATER,
}

ESCAPES = {'"': '"', "\\": "\\", "n": "\n", "r": "\r", "t": "\t"}


@dataclass(frozen=True)
class Token:
    kind: Kind
    text: str
    offset: int

    @property
    def end(self) -> int:
        return self.offset + len(self.text)

    def string_value(self) -> str:
        assert self.kind == Kind.STRING
        out = []
        i = 1
        while i < len(self.text) - 1:
            c = self.text[i]
            if c == "\\":
                out.append(ESCAPES[self.text[i + 1]])
                i += 2
            else:
                out.append(c)
                i += 1
        return "".join(out)


def position(source: str, offset: int) -> tuple[int, int]:
    """1-based (line, column) of an offset."""
    line = source.count("\n", 0, offset) + 1
    start = source.rfind("\n", 0, offset) + 1
    return line, offset - start + 1


class LexError(Exception):
    def __init__(self, message: str, offset: int):
        super().__init__(message)
        self.message = message
        self.offset = offset


def is_name_start(c: str) -> bool:
    return c.isalpha() or c == "_"


def is_name_continue(c: str) -> bool:
    return c.isalnum() or c == "_"


def lex(source: str) -> list[Token]:
    tokens = []
    i = 0
    n = len(source)
    while i < n:
        c = source[i]
        if c in " \t\r\n":
            i += 1
        elif c == "#":
            end = source.find("\n", i)
            i = n if end == -1 else end
        elif source[i : i + 2] in TWO_CHAR:
            tokens.append(Token(TWO_CHAR[source[i : i + 2]], source[i : i + 2], i))
            i += 2
        elif c in ONE_CHAR:
            tokens.append(Token(ONE_CHAR[c], c, i))
            i += 1
        elif is_name_start(c):
            j = i + 1
            while j < n and is_name_continue(source[j]):
                j += 1
            text = source[i:j]
            tokens.append(Token(KEYWORDS.get(text, Kind.NAME), text, i))
            i = j
        elif c == '"':
            j = i + 1
            while True:
                if j >= n or source[j] == "\n":
                    raise LexError("unclosed string", i)
                if source[j] == '"':
                    break
                if source[j] == "\\":
                    if j + 1 >= n or source[j + 1] not in ESCAPES:
                        raise LexError("invalid escape sequence", j)
                    j += 2
                else:
                    j += 1
            tokens.append(Token(Kind.STRING, source[i : j + 1], i))
            i = j + 1
        else:
            raise LexError(f"unexpected character {c!r}", i)
    tokens.append(Token(Kind.EOF, "", n))
    return tokens
